util/deephash/deephash.go

```diff
@@ -140,31 +153,36 @@ func (h *hasher) print(v reflect.Value) (acyclic bool) {
 		return h.print(v.Elem())
 	case reflect.Struct:
 		acyclic = true
-		w.WriteString("struct{\n")
+		w.WriteString("struct")
+		h.int(v.NumField())
 		for i, n := 0, v.NumField(); i < n; i++ {
-			fmt.Fprintf(w, " [%d]: ", i)
+			h.int(i)
 			if !h.print(v.Field(i)) {
 				acyclic = false
 			}
-			w.WriteString("\n")
 		}
-		w.WriteString("}\n")
 		return acyclic
 	case reflect.Slice, reflect.Array:
+		Len := v.Len()
+		if v.Kind() == reflect.Slice {
+			h.int(Len)
+		}
 		if v.Type().Elem().Kind() == reflect.Uint8 && v.CanInterface() {
-			fmt.Fprintf(w, "%q", v.Interface())
+			if Len > 0 && Len <= scratchSize {
+				n := reflect.Copy(reflect.ValueOf(&h.scratch).Elem(), v)
```

Comment:
Does ValueOf allocate? If not, yay, the compiler is getting better.

Reply:
It hasn't for a long time, as long as you don't allocate into the interface. Hence the &+Elem.